README
Ajax Handler v1.3 - CakePHP Component
*************************************

Requirements
============

    * CakePHP 1.2.x, 1.3.x
    * PHP 5.2.x, 5.3.x
    * SimpleXML

Features
========

    * Handles pre-defined Controller actions as AJAX
    * Formats the AJAX Post/Get into Controller $data values
    * Blackholes / kills the request if the action is not called through AJAX
    * Respond with a success or failure message
    * Automatically format your data into JSON, XML, HTML or plain text
    * Responds with the correct Content-Type headers
    * Utilizes the Request Handler and Type Converter

Documentation
=============

    * http://milesj.me/resources/script/ajax-handler-component


Docs/changelog.md
# Changelog #

*These logs may be outdated or incomplete.*

## 2.0.0 ##

* Updated to CakePHP 2.0 (not backwards compatible with 1.3)
* Used the Controller CakeRequest object in place of RequestHandler
* Renamed $__handledActions to $_handled and made protected
* Removed $__responseTypes in favor of CakeResponse
* Removed the data gathering from startup() in favor of CakeRequest
* Removed valid() as that should be done externally

## 1.6 ##

* Made operation comparisons strict
* Removed the 2nd argument of respond()
* Re-added the $this->Controller references

## 1.5 ##

* Changed $allowRemoteRequests to $allowRemote
* Removed $this->Controller references
* response() now takes an array as the 3rd argument
* Updated TypeConverter

## 1.4 ##

* JSON encoding now uses TypeConverter
* Updated TypeConverter version
* Removed PHP 4 argument references

## 1.3 ##

* Updated to CakePHP 1.3
* Added the TypeConverter class to handle the XML to array (and vice versa) conversion process
* Removed the response "message" key

## 1.2 ##

* Added an $allowRemoteRequests property to determine if remote AJAX calls should be allowed
* Moved the debug configuration to only apply for AJAX calls
* Added logic that will add a "messages" parameter to the response if a Flash message is set

## 1.1 ##

* Upgraded to PHP 5 only
* Added more security checking to make sure the AJAX call comes from the same domain
* Added functionality for objects within __format()

## 1.0 ##

* First initial release of the Ajax Handler component


vendors/TypeConverter.php
<?php
/**
 * A class that handles the detection and conversion of certain resource formats / content types into other formats.
 * The current formats are supported: XML, JSON, Array, Object, Serialized
 *
 * @author		Miles Johnson - http://milesj.me
 * @copyright	Copyright 2006-2010, Miles Johnson, Inc.
 * @license		http://opensource.org/licenses/mit-license.php - Licensed under The MIT License
 */

class TypeConverter {

	/**
	 * Current version.
	 *
	 * @access public
	 * @var string
	 */
	public static $version = '1.1';

	/**
	 * Disregard XML attributes and only return the value.
	 */
	const XML_NONE = 0;

	/**
	 * Merge attributes and the value into a single dimension; the values key will be "value".
	 */
	const XML_MERGE = 1;

	/**
	 * Group the attributes into a key "attributes" and the value into a key of "value".
	 */
	const XML_GROUP = 2;

	/**
	 * Attributes will only be returned.
	 */
	const XML_OVERWRITE = 3;

	/**
	 * Returns a string for the detected type.
	 *
	 * @access public
	 * @param mixed $data
	 * @return string
	 * @static
	 */
	public static function is($data) {
		if (self::isArray($data)) {
			return 'array';

		} else if (self::isObject($data)) {
			return 'object';

		} else if (self::isJson($data)) {
			return 'json';

		} else if (self::isSerialized($data)) {
			return 'serialized';

		} else if (self::isXml($data)) {
			return 'xml';
		}

		return 'other';
	}

	/**
	 * Check to see if data passed is an array.
	 *
	 * @access public
	 * @param mixed $data
	 * @return boolean
	 * @static
	 */
	public static function isArray($data) {
		return is_array($data);
	}

	/**
	 * Check to see if data passed is a JSON object.
	 *
	 * @access public
	 * @param mixed $data
	 * @return boolean
	 * @static
	 */
	public static function isJson($data) {
		return (@json_decode($data) !== null);
	}

	/**
	 * Check to see if data passed is an object.
	 *
	 * @access public
	 * @param mixed $data
	 * @return boolean
	 * @static
	 */
	public static function isObject($data) {
		return is_object($data);
	}

	/**
	 * Check to see if data passed has been serialized.
	 *
	 * @access public
	 * @param mixed $data
	 * @return boolean
	 * @static
	 */
	public static function isSerialized($data) {
		$ser = @unserialize($data);
		return ($ser !== false) ? $ser : false;
	}

	/**
	 * Check to see if data passed is an XML document.
	 *
	 * @access public
	 * @param mixed $data
	 * @return boolean
	 * @static
	 */
	public static function isXml($data) {
		$xml = @simplexml_load_string($data);
		return ($xml instanceof SimpleXmlElement) ? $xml : false;
	}

	/**
	 * Transforms a resource into an array.
	 *
	 * @access public
	 * @param mixed $resource
	 * @return array
	 * @static
	 */
	public static function toArray($resource) {
		if (self::isArray($resource)) {
			return $resource;

		} else if (self::isObject($resource)) {
			return self::buildArray($resource);

		} else if (self::isJson($resource)) {
			return json_decode($resource, true);

		} else if ($ser = self::isSerialized($resource)) {
			return self::toArray($ser);

		} else if ($xml = self::isXml($resource)) {
			return self::xmlToArray($xml);
		}

		return $resource;
	}

	/**
	 * Transforms a resource into a JSON object.
	 *
	 * @access public
	 * @param mixed $resource
	 * @return json
	 * @static
	 */
	public static function toJson($resource) {
		if (self::isJson($resource)) {
			return $resource;

		} else {
			if ($xml = self::isXml($resource)) {
				$resource = self::xmlToArray($xml);

			} else if ($ser = self::isSerialized($resource)) {
				$resource = $ser;
			}

			return json_encode($resource);
		}

		return $resource;
	}

	/**
	 * Transforms a resource into an object.
	 *
	 * @access public
	 * @param mixed $resource
	 * @return object
	 * @static
	 */
	public static function toObject($resource) {
		if (self::isObject($resource)) {
			return $resource;

		} else if (self::isArray($resource)) {
			return self::buildObject($resource);

		} else if (self::isJson($resource)) {
			return json_decode($resource);

		} else if ($ser = self::isSerialized($resource)) {
			return self::toObject($ser);

		} else if ($xml = self::isXml($resource)) {
			return $xml;
		}

		return $resource;
	}

	/**
	 * Transforms a resource into a serialized form.
	 *
	 * @access public
	 * @param mixed $resource
	 * @return string
	 * @static
	 */
	public static function toSerialize($resource) {
		if (!self::isArray($resource)) {
			$resource = self::toArray($resource);
		}

		return serialize($resource);
	}

	/**
	 * Transforms a resource into an XML document.
	 *
	 * @access public
	 * @param mixed $resource
	 * @param string $root
	 * @return string
	 * @static
	 */
	public static function toXml($resource, $root = 'root') {
		if (self::isXml($resource)) {
			return $resource;
		}

		$array = self::toArray($resource);

		if (!empty($array)) {
			$xml = simplexml_load_string('<?xml version="1.0" encoding="utf-8"?><'. $root .'></'. $root .'>');
			$response = self::buildXml($xml, $array);

			return $response->asXML();
		}

		return $resource;
	}

	/**
	 * Turn an object into an array. Alternative to array_map magic.
	 *
	 * @access public
	 * @param object $object
	 * @return array
	 */
	public static function buildArray($object) {
		$array = array();

		foreach ($object as $key => $value) {
			if (is_object($value)) {
				$array[$key] = self::buildArray($value);
			} else {
				$array[$key] = $value;
			}
		}

		return $array;
	}

	/**
	 * Turn an array into an object. Alternative to array_map magic.
	 *
	 * @access public
	 * @param array $array
	 * @return object
	 */
	public static function buildObject($array) {
		$obj = new stdClass();

		foreach ($array as $key => $value) {
			if (is_array($value)) {
				$obj->{$key} = self::buildObject($value);
			} else {
				$obj->{$key} = $value;
			}
		}

		return $obj;
	}

	/**
	 * Turn an array into an XML document. Alternative to array_map magic.
	 *
	 * @access public
	 * @param object $xml
	 * @param array $array
	 * @return object
	 */
	public static function buildXml(&$xml, $array) {
		if (is_array($array)) {
			foreach ($array as $key => $value) {
				if (is_array($value)) {

					// Multiple nodes of the same name
					if (isset($value[0])) {
						foreach ($value as $kValue) {
							if (is_array($kValue)) {
								self::buildXml($xml, array($key => $kValue));
							} else {
								$xml->addChild($key, $kValue);
							}
						}

					// XML_GROUP
					} else if (isset($value['attributes'])) {
						$node = $xml->addChild($key, $value['value']);

						foreach ($value['attributes'] as $aKey => $aValue) {
							$node->addAttribute($aKey, $aValue);
						}

					// XML_MERGE
					// XML_OVERWRITE
					} else {
						$node = $xml->addChild($key, (isset($value['value']) ? $value['value'] : ''));
						unset($value['value']);

						foreach ($value as $aKey => $aValue) {
							if (is_array($aValue)) {
								self::buildXml($node, array($aKey => $aValue));
							} else {
								$node->addAttribute($aKey, $aValue);
							}
						}
					}

				// XML_NONE
				} else {
					$xml->addChild($key, $value);
				}
			}
		}

		return $xml;
	}

	/**
	 * Convert a SimpleXML object into an array.
	 *
	 * @access public
	 * @param object $xml
	 * @param int $format
	 * @return array
	 */
	public static function xmlToArray($xml, $format = self::XML_GROUP) {
		if (is_string($xml)) {
			$xml = @simplexml_load_string($xml);
		}

		if ($xml->count() <= 0) {
			return (string)$xml;
		}

		$array = array();

		foreach ($xml->children() as $element => $node) {
			$data = array();

			if (!isset($array[$element])) {
				$array[$element] = "";
			}

			if (!$node->attributes() || $format == self::XML_NONE) {
				$data = self::xmlToArray($node, $format);

			} else {
				switch ($format) {
					case self::XML_GROUP:
						$data = array(
							'attributes' => array(),
							'value' => (string)$node
						);

						foreach ($node->attributes() as $attr => $value) {
							$data['attributes'][$attr] = (string)$value;
						}
					break;

					case self::XML_MERGE:
					case self::XML_OVERWRITE:
						foreach ($node->attributes() as $attr => $value) {
							$data[$attr] = (string)$value;
						}

						if ($format == self::XML_MERGE) {
							$data['value'] = (string)$node;
						}
					break;
				}
			}

			if (count($xml->{$element}) > 1) {
				$array[$element][] = $data;
			} else {
				$array[$element] = $data;
			}
		}

		return $array;
	}

}


Controller/Component/AjaxHandlerComponent.php
<?php
/** 
 * Ajax Handler Component
 *
 * A CakePHP Component that will automatically handle and render AJAX calls and apply the appropriate returned format and headers.
 *
 * @author      Miles Johnson - http://milesj.me
 * @copyright   Copyright 2006-2011, Miles Johnson, Inc.
 * @license     http://opensource.org/licenses/mit-license.php - Licensed under The MIT License
 * @link        http://milesj.me/code/cakephp/ajax-handler
 */

App::import('Vendor', 'TypeConverter');

class AjaxHandlerComponent extends Component {

	/**
	 * Current version.
	 *
	 * @access public
	 * @var string
	 */
	public $version = '2.0.1';
	
	/**
	 * Components.
	 *
	 * @access public
	 * @var array
	 */
	public $components = array('RequestHandler');

	/**
	 * Should we allow remote AJAX calls.
	 *
	 * @access public
	 * @var boolean
	 */
	public $allowRemote = false;

	/**
	 * A user given code associated with failure / success messages.
	 *
	 * @access protected
	 * @var int
	 */
	protected $_code;

	/**
	 * Contains the success messages / errors.
	 *
	 * @access protected
	 * @var array
	 */
	protected $_data;

	/**
	 * Which actions are handled as AJAX.
	 *
	 * @access protected
	 * @var array
	 */
	protected $_handled = array();

	/**
	 * Determines if the AJAX call was a success or failure.
	 *
	 * @access protected
	 * @var boolean
	 */
	protected $_success = false;

	/**
	 * Load the Controller object.
	 *
	 * @access public
	 * @param Controller $controller
	 * @return void
	 */
	public function initialize($controller) {
		if ($controller->request->is('ajax')) {
			Configure::write('debug', 0);

			// Must disable security component for AJAX
			if (isset($controller->Security)) {
				$controller->Security->validatePost = false;
				$controller->Security->csrfCheck = false;
			}

			// If not from this domain, destroy
			if (($this->allowRemote === false) && (strpos(env('HTTP_REFERER'), trim(env('HTTP_HOST'), '/')) === false)) {
				if (isset($controller->Security)) {
					$controller->Security->blackHole($controller, 'Invalid referrer detected for this request.');
				} else {
					$controller->redirect(null, 403, true);
				}
			}
		}

		$this->controller = $controller;
	}

	/**
	 * Determine if the action is an AJAX action and handle it.
	 *
	 * @access public
	 * @param Controller $controller
	 * @return void
	 */
	public function startup($controller) {
		$handled = ($this->_handled === array('*') || in_array($controller->action, $this->_handled));

		if ($handled && !$controller->request->is('ajax')) {
			if (isset($controller->Security)) {
				$controller->Security->blackHole($controller, 'You are not authorized to process this request.');
			} else {
				$controller->redirect(null, 401, true);
			}
		}

		$this->controller = $controller;
	}

	/**
	 * A list of actions that are handled as an AJAX call.
	 *
	 * @access public
	 * @return void
	 */
	public function handle() {
		$actions = func_get_args();

		if ($actions === array('*') || empty($actions)) {
			$this->_handled = array('*');
		} else {
			$this->_handled = array_unique(array_intersect($actions, get_class_methods($this->controller)));
		}
	}

	/**
	 * Respond the AJAX call with the gathered data.
	 *
	 * @access public
	 * @param string $type
	 * @param array $response
	 * @return mixed
	 */
	public function respond($type = 'json', array $response = array()) {
		if (!empty($response)) {
			$response = $response + array(
				'success' => false,
				'data' => '',
				'code' => null
			);

			$this->response($response['success'], $response['data'], $response['code']);
		}

		if ($type == 'html') {
			$this->RequestHandler->renderAs($this->controller, 'ajax');

		} else {
			$this->RequestHandler->respondAs($type);
			$this->controller->autoLayout = false;
			$this->controller->autoRender = false;

			echo $this->__format($type);
		}
	}

	/**
	 * Handle the response as a success or failure alongside a message or error.
	 *
	 * @access public
	 * @param boolean $success
	 * @param mixed $data
	 * @param mixed $code
	 * @return void
	 */
	public function response($success, $data = '', $code = null) {
		$this->_success = (bool) $success;
		$this->_data = $data;
		$this->_code = $code;
	}
	
	/**
	 * What should happen if the class is called stand alone.
	 *
	 * @access public
	 * @return mixed
	 */
	public function __toString() {
		return $this->respond();
	}

	/**
	 * Format the response into the right content type.
	 *
	 * @access private
	 * @param string $type
	 * @return mixed
	 */
	private function __format($type) {
		$response = array(
			'success' => $this->_success,
			'data' => $this->_data
		);

		if (!empty($this->_code)) {
			$response['code'] = $this->_code;
		}

		switch (strtolower($type)) {
			case 'json':
				$format = TypeConverter::toJson($response);
			break;
			case 'xml':
				$format = TypeConverter::toXml($response);
			break;
			case 'html';
			case 'text':
			default:
				$format = (string) $this->_data;
			break;
		}

		return $format;
	}

}


controllers/components/ajax_handler.php
<?php
/** 
 * Ajax Handler Component
 *
 * A CakePHP Component that will automatically handle and render AJAX calls and apply the appropriate returned format and headers.
 *
 * @author      Miles Johnson - http://milesj.me
 * @copyright   Copyright 2006-2011, Miles Johnson, Inc.
 * @license     http://opensource.org/licenses/mit-license.php - Licensed under The MIT License
 * @link        http://milesj.me/resources/script/ajax-handler-component
 */

App::import(array(
	'type' => 'Vendor',
	'name' => 'TypeConverter',
	'file' => 'TypeConverter.php'
));
 
class AjaxHandlerComponent extends Object {

	/**
	 * Current version: http://milesj.me/resources/logs/ajax-handler-component
	 *
	 * @access public
	 * @var string
	 */
	public $version = '1.6';

	/**
	 * Components.
	 *
	 * @access public
	 * @var array
	 */
	public $components = array('RequestHandler');

	/**
	 * Should we allow remote AJAX calls.
	 *
	 * @access public
	 * @var boolean
	 */
	public $allowRemote = false;

	/**
	 * Determines if the AJAX call was a success or failure.
	 *
	 * @access protected
	 * @var boolean
	 */
	protected $_success = false;

	/**
	 * A user given code associated with failure / success messages.
	 *
	 * @access protected
	 * @var int
	 */
	protected $_code;

	/**
	 * Contains the success messages / errors.
	 *
	 * @access protected
	 * @var array
	 */
	protected $_data;

	/**
	 * Which actions are handled as AJAX.
	 *
	 * @access protected
	 * @var array
	 */
	private $__handledActions = array();

	/**
	 * Types to respond as.
	 *
	 * @access protected
	 * @var array
	 */
	private $__responseTypes = array(
		'json'	=> 'application/json',
		'html'	=> 'text/html',
		'xml'	=> 'text/xml',
		'text'	=> 'text/plain'
	);

	/**
	 * Load the Controller object.
	 *
	 * @access public
	 * @param object $Controller
	 * @return void
	 */
	public function initialize($Controller) {
		if ($this->RequestHandler->isAjax()) {
			if (isset($this->allowRemoteRequests)) {
				$this->allowRemote = $this->allowRemoteRequests;
			}

			// Turn off debug, don't want to ruin our response
			Configure::write('debug', 0);

			// Must disable security component for AJAX
			if (isset($Controller->Security)) {
				$Controller->Security->validatePost = false;
			}

			// If not from this domain, destroy
			if (($this->allowRemote === false) && (strpos(env('HTTP_REFERER'), trim(env('HTTP_HOST'), '/')) === false)) {
				if (isset($Controller->Security)) {
					$Controller->Security->blackHole($Controller, 'Invalid referrer detected for this request!');
				} else {
					$Controller->redirect(null, 403, true);
				}
			}
		}

		$this->Controller = $Controller;
	}

	/**
	 * Determine if the action is an AJAX action and handle it.
	 *
	 * @access public
	 * @param object $Controller
	 * @return void
	 */
	public function startup($Controller) {
		$handled = false;

		if ($this->__handledActions === array('*') || in_array($Controller->action, $this->__handledActions)) {
			$handled = true;
		}

		if (!$this->RequestHandler->isAjax() && $handled) {
			if (isset($Controller->Security)) {
				$Controller->Security->blackHole($Controller, 'You are not authorized to process this request!');
			} else {
				$Controller->redirect(null, 401, true);
			}
		}

		// Load up the controller with data
		if ($handled) {
			$data = array();

			if (!empty($Controller->params['form'])) {
				$data = $Controller->params['form'] + $data;
			}

			if (!empty($Controller->params['url'])) {
				$data = $Controller->params['url'] + $data;
				unset($data['ext'], $data['url']);
			}

			if (!empty($data)) {
				$data = array_map('urldecode', $data);

				if (!empty($Controller->data)) {
					$Controller->data = $data + $Controller->data;
				} else {
					$Controller->data = $data;
				}
			}
		}

		$this->Controller = $Controller;
	}

	/**
	 * A list of actions that are handled as an AJAX call.
	 *
	 * @access public
	 * @return void
	 */
	public function handle() {
		$actions = func_get_args();

		if ($actions === array('*') || empty($actions)) {
			$this->__handledActions = array('*');

		} else if (is_array($actions) && !empty($actions)) {
			$this->__handledActions = array_unique(array_intersect($actions, get_class_methods($this->Controller)));
		}
	}

	/**
	 * Respond the AJAX call with the gathered data.
	 *
	 * @access public
	 * @param string $type
	 * @param array $response
	 * @return mixed
	 */
	public function respond($type = 'json', $response = array()) {
		if (empty($this->__responseTypes[$type])) {
			$type = 'json';
		}

		// Apply response
		if (!empty($response) && is_array($response)) {
			$response = $response + array(
				'success' => null,
				'data' => '',
				'code' => null
			);

			$this->response($response['success'], $response['data'], $response['code']);
		}

		// Set to null for Cake 1.2
		$this->RequestHandler->__responseTypeSet = null;

		if ($type == 'html') {
			$this->RequestHandler->renderAs($this->Controller, 'ajax');
			$this->Controller->autoLayout = true;
			$this->Controller->autoRender = true;

		} else {
			$this->RequestHandler->respondAs($this->__responseTypes[$type]);
			$this->Controller->autoLayout = false;
			$this->Controller->autoRender = false;

			echo $this->__format($type);
		}
	}

	/**
	 * Handle the response as a success or failure alongside a message or error.
	 *
	 * @access public
	 * @param boolean $success
	 * @param mixed $data
	 * @param mixed $code
	 * @return void
	 */
	public function response($success, $data = '', $code = null) {
		if (is_bool($success)) {
			$this->_success = $success;
		}

		$this->_data = $data;
		$this->_code = $code;
	}

	/**
	 * Makes sure the params passed are clean.
	 *
	 * @access public
	 * @param string|int $request
	 * @param boolean $isString
	 * @return mixed
	 */
	public function valid($request, $isString = false) {
		if ($isString) {
			return (isset($request) && is_string($request) && $request != '');
		} else {
			return (isset($request) && is_numeric($request));
		}
	}

	/**
	 * What should happen if the class is called stand alone.
	 *
	 * @access public
	 * @return mixed
	 */
	public function __toString() {
		return $this->respond();
	}

	/**
	 * Format the response into the right content type.
	 *
	 * @access private
	 * @param string $type
	 * @return mixed
	 */
	private function __format($type) {
		$response = array(
			'success' => $this->_success,
			'data' => $this->_data
		);

		if (!empty($this->_code)) {
			$response['code'] = $this->_code;
		}

		switch (strtolower($type)) {
			case 'json':
				$format = TypeConverter::toJson($response);
			break;
			case 'xml':
				$format = TypeConverter::toXml($response);
			break;
			case 'html';
			case 'text':
			default:
				$format = (string)$this->_data;
			break;
		}

		return $format;
	}

}
